cmd/snap-update-ns: add C preamble for setns

zyga opened the pull request:
This patch adds a preamble/bootstrap code to snap-update-ns so that it
can call setns(..., CLONE_NEWNS) before golang starts to create threads.

The setns(2) manual page states that CLONE_NEWNS fails if there are any
threads present, apart from the main thread.

The preamble has no access to command line arguments so it works around
this by probing /proc/self/cmdline and parsing this manually.

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

niemeyer commented on cmd/snap-update-ns/bootstrap.c:
```diff
@@ -0,0 +1,129 @@
+    if (argv0_len == num_read) {
+        return NULL;
+    }
+    return &buf[argv0_len + 1];
```
If \0 is the last byte, then argv_len < num_read, yet argv0_len+1 is an overrun.

zyga replied:
Corrected

niemeyer commented on cmd/snap-update-ns/bootstrap.c:
```diff
@@ -0,0 +1,129 @@
+    }
+    memset(buf, 0, buf_size);
+    ssize_t num_read;
+    if ((num_read = read(fd, buf, buf_size)) < 0) {
```
More readable to have the read outside the if. Then, ideally this would also check:
```
if (num_read < 0 || num_read == sizeof buf_size)

zyga replied:
Tweaked

niemeyer commented on cmd/snap-update-ns/bootstrap.c:
```diff
@@ -0,0 +1,129 @@
+    // Find the name of the snap by scanning the cmdline.  If there's no snap
+    // name given, just bail out. The go parts will scan this too.
+    const char* snap_name = find_snap_name(cmdline, sizeof cmdline, num_read);
+    if (*snap_name == '\0') {
```
find_snap_name returns NULL.

Slightly worried about the rate of important errors in this logic. Please make sure to get a good second review.

zyga replied:
Ack, thank you for the first review, will do!

morphis commented on cmd/snap-update-ns/bootstrap.h:
```diff
@@ -0,0 +1,20 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
```
Please add proper header guards

#ifndef SNAP_UPDATE_NS_BOOTSTRAP_H_
#define SNAP_UPDATE_NS_BOOTSTRAP_H_
[...]
#endif

zyga replied:
Ack, will do

zyga replied:
Done

morphis commented on cmd/snap-update-ns/bootstrap.h:
```diff
@@ -0,0 +1,20 @@
+ *
+ */
+
+extern int bootstrap_errno;
```
This one is only read from other code parts so a _get function is more what we want here.

zyga replied:
I decided not to, it's just more code that needs to be written where we want less code to exist.

morphis commented on cmd/snap-update-ns/bootstrap.h:
```diff
@@ -0,0 +1,20 @@
+ */
+
+extern int bootstrap_errno;
+extern const char* bootstrap_msg;
```
Same as for bootstrap_errno

zyga replied:
Not sure if this buys us anything. Would make the C code longer.

morphis replied:
Would just make the API contract clear. We don't support write, just read.

morphis commented on cmd/snap-update-ns/bootstrap.c:
```diff
@@ -0,0 +1,152 @@
+ *
+ */
+
+#include "bootstrap.h"
```
Not really needed.

zyga replied:
Why not? This is standard practice to ensure that headers match implementation.

morphis replied:
Ah yeah, we're in C and not in C++, so makes more sense here to spot API/implementation differences.

jdstrand commented:
I'm struggling to see how update-ns being in Go is really saving us on complexity and maintenance when I see comments like this:

```
// Use a pre-main helper to switch the mount namespace. This is required as
// golang creates threads at will and setns(..., CLONE_NEWNS) fails if any
// threads apart from the main thread exist.
```

which is making us have to write C code to do the OS-level operations that are required.

jdstrand commented:
In general, cmdline is something that is user-controlled. I understand that snapd is the caller of update-ns, but how are the arguments to update-ns created? Is it possible for a user to invoke a snap run command and manipulate cmdline? If so, this preamble code needs to be carefully written to handled untrusted input via cmdline.

zyga commented:
@jdstrand to reply to your concerns:

The amount of system C code we need to write is minimal, it's really just one system call. The only complication is that we have no access to the command line in a convenient way and must execute this function before golang bootstrap begins, at which we no longer have one thread.

This program is not setuid root and does not process user input in C. The golang parts, which are significantly harder to attack, process content generated by snapd (the mount profile) based on what two participating snaps declare. Having said that at this level there is no user input parsed.

mvo5 commented on cmd/snap-update-ns/bootstrap.c:
```diff
@@ -0,0 +1,149 @@
+    }
+    memset(buf, 0, buf_size);
+    ssize_t num_read = read(fd, buf, buf_size);
+    if (num_read < 0 || num_read == buf_size) {
```
Actually - I'm slightly confused. In the case of `num_read == buf_size` the bootstrap_{errno,msg} are set. However in the later code its only an error if num_read is < 0 - is there code missing here that sets num_read = -1 if num_read == buf_size ? Or am I just misreading this code altogether :) ?

zyga replied:
Thanks, I'll tweak this accordingly. I think you are right.

mvo5 commented on cmd/snap-update-ns/bootstrap.c:
```diff
@@ -0,0 +1,149 @@
+
+// find_snap_name scans the command line buffer and looks for the 1st argument.
+const char*
+find_snap_name(char* buf, size_t buf_size, size_t num_read)
```
Silly(?) question - but why do we need both buf_size and num_read here? Isn't num_read always smaller than buf_size and sufficient for the size checks in here?

zyga replied:
Just for simplicity and correctness, `buf_size` is a constant.

mvo5 replied:
Honest question, why is it simpler (and more correct) to use `buf_size` for the first check and num_read for the second? I mean, it seems checking for just num_read in the strnlen() is equally correct (and arguably simpler because when reading the code there is no need to ponder why in one place buf_size is used and in the other place num_read. Don't get me wrong, its fine, I'm just trying to understand the background a bit better.

zyga replied:
I think it is simpler because one argument is not a variable but a compile-time constant. I can fuse this into one variable if you want but I like the explicitness of "this is the size of the buffer" in C APIs.

niemeyer replied:
Having more logic to think about makes things more complex rather than simpler. If we should never go above the smaller number &ndash; and we shouldn't &ndash; there's no reason to have another number and make things any more complex.

Trivial and potentially serious (tomorrow) bug to introduce: invert the two parameters.

mvo5 commented on cmd/snap-update-ns/bootstrap.c:
```diff
@@ -0,0 +1,149 @@
+        bootstrap_errno = errno;
+        bootstrap_msg = "cannot read /proc/self/cmdline";
+    }
+    close(fd);
```
Close may also return an error

zyga replied:
In practice it is an useless error that should not be checked for or handled. See: https://lwn.net/Articles/576478/

mvo5 commented on cmd/snap-update-ns/main.go:
```diff
@@ -42,13 +44,19 @@ func main() {
 	parser := flags.NewParser(&opts, flags.HelpFlag|flags.PassDoubleDash|flags.PassAfterNonOption)
 	_, err := parser.ParseArgs(args)
-	return err
+	if err != nil {
```
This could be written as a single line: `if _, err := parser.ParseArgs(args); err != nil {...`

zyga replied:
Done

mvo5 commented on cmd/snap-update-ns/main.go:
```diff
@@ -42,13 +44,19 @@ func main() {
 		return err
 	}
-
+	if err := bootstrapError(); err != nil {
```
Maybe a small comment here that that the bootstrap C code is automatically called and that we need to check errors for that here?

zyga replied:
Will do, thanks

zyga replied:
Done

zyga commented:
I updated this, tweaking slightly how initialization is done so that we can test more code. I added unit tests for what is reasonable to test this way (skipping the actual setns as that will be covered by spread). I can add a smoke spread test that just checks we setns correctly and reach the "not implemented" trigger (I will do that in a moment) but I'd like to land this and iterate.

mvo5 commented on cmd/snap-update-ns/bootstrap.c:
```diff
@@ -0,0 +1,165 @@
+        bootstrap_msg = "cannot read /proc/self/cmdline";
+    } else if (num_read == buf_size) {
+        bootstrap_errno = 0;
+        bootstrap_msg = "cannot fit all of /proc/self/cmdline, buffer too small";
```
I think this also needs to set `num_read = -1` or else read_cmdline() will not fail in https://github.com/snapcore/snapd/pull/3095/files#diff-018c9af8a9874391aef7a2bfa0300d19R142 but carry on with an incorrect buffer.

mvo5 commented on cmd/snap-update-ns/bootstrap.c:
```diff
@@ -0,0 +1,165 @@
+// on command line.
+void bootstrap(void)
+{
+// NOTE: This lets use use cgo/go to write tests without running the bulk
```
Typo: `This lets *us* use`...

zyga commented:
@mvo5 (I cannot answer your question up there for some reason). I think it is simpler because one argument is not a variable but a compile-time constant. I can fuse this into one variable if you want but I like the explicitness of "this is the size of the buffer" in C APIs.

stolowski commented on cmd/snap-update-ns/bootstrap.c:
```diff
@@ -0,0 +1,166 @@
+
+// sanitize_snap_name performs partial validation of the given name.
+// The goal is to ensure that there are no / or .. in the name.
+int sanitize_snap_name(const char* snap_name)
```
Nitpick: wouldn't "validate" be a better name since this method doesn't actually alter the name in any way?

zyga replied:
Yes, good point. How about `partially_validate_snap_name` (to make it clear it's not a full validation)?

stolowski commented on cmd/snap-update-ns/bootstrap.c:
```diff
@@ -0,0 +1,166 @@
+// of the code automatically. In snapd we can just set the required
+// environment variable.
+#define TRIGGER_KEY "SNAPD_INTERNAL"
+#define TRIGGER_VAL "x-switch-namespace=1,"
```
Is x-... some form of new convention we want to introduce with this PR? Do we use it elsewhere in snapd? Looks a little verbose.

zyga replied:
Ha, I was just trying to be generic, the`SNAPD_INTERNAL` variable could hold a list of comma-separated key-value pairs. I'm fine with reworking this in any way that we find suitable.

stolowski replied:
I see. Ok, no arguing with that, just wanted to understand the reason ;)

niemeyer replied:
Something more straightforward would be nice indeed. How about
```
SNAPD_DEBUG_UPDATE_NS="skip-bootstrap"

pedronis replied:
why not just have a bit of C code that sets a global no_bootstrap flag, that is compiled in only when building tests?

zyga replied:
Because cgo cannot be used in tests.

pedronis replied:
you can use a file with a build tag though, no?

niemeyer commented on cmd/snap-update-ns/bootstrap.c:
```diff
@@ -0,0 +1,166 @@
+    // Find the name of the snap by scanning the cmdline.  If there's no snap
+    // name given, just bail out. The go parts will scan this too.
+    const char* snap_name = find_snap_name(cmdline, sizeof cmdline, num_read);
+    if (snap_name == NULL || *snap_name == '\0') {
```
The second case would be more safely handled inside find_snap_name. If it got an empty string, it didn't actually find a snap name.

zyga replied:
+1

niemeyer commented on cmd/snap-update-ns/bootstrap.c:
```diff
@@ -0,0 +1,166 @@
+    // Look for known offenders in the snap name (.. and /) and do nothing if
+    // those are found. The golang code will validate snap name and print a
+    // proper error message but this just ensures we don't try to open / setns
+    // anything unusual.
```
Nice.

niemeyer commented on cmd/snap-update-ns/bootstrap.go:
```diff
@@ -0,0 +1,77 @@
+// findSnapName parses the argv-like array and finds the 1st argument.
+// Subsequent arguments are spearated by NUL-bytes.
+func findSnapName(buf []byte) *string {
+	if ptr := C.find_snap_name((*C.char)(unsafe.Pointer(&buf[0])), C.size_t(cap(buf)), C.size_t(len(buf))); ptr != nil {
```
That doesn't look safe. That call is returning with ptr pointing to memory managed by Go's GC. If the memory moves, the function below will be accessing unknown memory. Instead, buf should be copied out into a malloc controlled buffer, defer free, and then passed into find_snap_name.

zyga replied:
+1

zyga replied:
On 2nd look I think this is correct:
```
 The major change is the definition of rules for sharing Go pointers with C code, to ensure that such C code can coexist with Go's garbage collector. Briefly, Go and C may share memory allocated by Go when a pointer to that memory is passed to C as part of a cgo call, provided that the memory itself contains no pointers to Go-allocated memory, and provided that C does not retain the pointer after the call returns. These rules are checked by the runtime during program execution: if the runtime detects a violation, it prints a diagnosis and crashes the program. The checks can be disabled by setting the environment variable GODEBUG=cgocheck=0, but note that the vast majority of code identified by the checks is subtly incompatible with garbage collection in one way or another. Disabling the checks will typically only lead to more mysterious failure modes. Fixing the code in question should be strongly preferred over turning off the checks. See the cgo documentation for more details. 
```
This is from cgo 1.6 release notes

niemeyer replied:
What I said above still seems true.

niemeyer replied:
Sorry, I think you are right. I was looking at the result of the function as if it was data that wouldn't be moved, but it's already a pointer, so if there's a move the GC will take it along as well.

niemeyer commented on cmd/snap-update-ns/bootstrap.go:
```diff
@@ -0,0 +1,77 @@
+func findSnapName(buf []byte) *string {
+	if ptr := C.find_snap_name((*C.char)(unsafe.Pointer(&buf[0])), C.size_t(cap(buf)), C.size_t(len(buf))); ptr != nil {
+		str := C.GoString(ptr)
+		return &str
```
`*string` is pretty unusual in Go. If we're wrapping, we can use Go's convention and return (string, error).

zyga replied:
(this is just for unit testing btw) I did it to be able to differentiate `nil` and `""` as returned from `find_snap_name`. I wanted to avoid extra wrappers as the real intent is to test the C implementation, not provide a Go-like API for something that is not used in Go otherwise.

niemeyer replied:
You have extra wrappers already. It is Go code that is deciding when to return nil or &str.

niemeyer commented on cmd/snap-update-ns/bootstrap.c:
```diff
@@ -0,0 +1,166 @@
+#define TRIGGER_KEY "SNAPD_INTERNAL"
+#define TRIGGER_VAL "x-switch-namespace=1,"
+    const char* snapd_internal = getenv(TRIGGER_KEY);
+    if (snapd_internal == NULL || strstr(snapd_internal, TRIGGER_VAL) == NULL) {
```
The logic here looks reversed. Bootstraping is the normal case, so the variable should be used to disable it instead, right?

zyga replied:
The problem is that we cannot say "don't bootstrap" from go. By the time go runs it's too late. I think it has to stay reversed.

niemeyer replied:
We don't want update-ns to require setting a debug flag to work at all, in the normal case.

zyga commented:
Thank you for the review. I'll address the trivial bits ASAP but I think the fundamental question about how when to skip the bootstrap code needs to stay as-is. The problem is, as I mentioned above, that we cannot do anything from go to say "don't bootstrap" because the whole bootstrap code runs before any go starts. We need to not bootstrap by default and let snapd set something that would make the bootstrap happen. I made sure we don't miss this by always failing if the bootstrap code is skipped so that in production if we skip bootstrap (for whatever reason) we will fail without doing anything else.

niemeyer commented:
Think outside the box. What hints you can use to decide not to bootstrap?

zyga commented:
@niemeyer I've updated the bootstrap logic to look at argv0 and skip bootstrap automatically if it ends with `".test"`. There's no more `SNAPD_INTERNAL` hack needed.

niemeyer commented on cmd/snap-update-ns/bootstrap.c:
```diff
@@ -0,0 +1,187 @@
+        return NULL;
+    }
+    char* snap_name = &buf[argv0_len + 1];
+    if (snap_name != NULL && *snap_name == '\0') {
```
It can't possibly be NULL here.

zyga replied:
aha, good catch :-) I'll remove that and merge! :)
